add support for package extensions (#3264)
JuliaLang/Pkg.jl: docs/src/creating-packages.md

Every dependency should in general have a compatibility constraint on it.
This is an important topic so there is a separate chapter about it: [Compatibility](@ref Compatibility).

## Conditional loading of code in packages (Glue packages)

!!! note
    This is a somewhat advanced section which can be skipped for people new to Julia and Julia packages.

It is sometimes desirable to be able to extend some functionality of a package without having to
unconditionally take on the cost (in terms of e.g. load time) of adding an extra dependency.
A *glue package* is a file that gets automatically loaded when some other set of packages are
loaded into the Julia session. This is very similar to functionality that the external package

> Requires.jl used to provide, but which is now aviable directly through Julia.

Reviewer: ```suggestion
Requires.jl used to provide, but which is now available directly through Julia.
```

A useful application of glue packages could be for a plotting package that should be able to plot
objects from a wide variety of different Julia packages.
Adding all those different Julia packages as dependencies
could be expensive since they would end up getting loaded even if they were never used.

> Instead, these code required to plot objects for specific packages can be put into separate files

Reviewer 2: ```suggestion
Instead, the code required to plot objects for specific packages can be put into separate files
```

> (glue packages) which is only loaded when 

Reviewer 2: unfinished sentence?

Below is an example of how the code can be structured for a use case as outlined above.

 `Project.toml`:
 ```toml
name = "Plotting"
version = "0.1.0"
uuid = "..."

[gluedeps]
Contour = "d38c429a-6771-53c6-b99e-75d170b6e991"

[gluepkgs] 
# name of glue package to the left
# glue dependencies required to load the glue pkg to the right
# use a list for multiple glue dependencies
GlueContour = "Contour" 

[compat] # compat can also be given on glue dependencies
Contour = "0.6.2"
```

`src/Plotting.jl`:
```julia
module Plotting

function plot(x::Vector)
    # Some functionality for plotting a vector here
end

end # module
```

`glue/GlueContour.jl` (can also be in `glue/GlueContour/GlueContour.jl`):
```julia
module GlueContour # Should be same name as the file (just like a normal package)

using Plotting, Contour

function Plotting.plot(c::Contour.ContourCollection)
    # Some functionality for plotting a contour here
end

end # module
```

A user that depends on `Plotting` will not pay the cost of the "glue code" inside the `GlueContour` module.
It is only when the `Contour` package actually gets loaded that the `GlueCountour` glue package is loaded

> and provide the new functionality.

Reviewer 2: ```suggestion
and provides the new functionality.
```

target to make them available when testing.

A glue package will only be loaded if the glue dependencies are loaded from the same environment or environments higher in the environment stack than the package itself.

!!! compat
    In order to be compatible with earlier versions of Julia, some extra steps have to be taken.
    These are: Duplicate the packages under `[gluedeps]` into `[extras]`. This is an unfortunate
    duplication but without doing this the project verifier under older Julia versions will complain (error).

### Backwards compatibility with Requires.jl

Since glue packages and Requires.jl are solving the same problem,
it is useful to know how to use Requires.jl on older Julia versions while seamlessly
transitioning into using glue packages on newer Julia versions that support it.
This is done by making the following changes (using the example above):

- Add the following to the package file. This makes it so that Requires.jl loads and inserts the
  callback only when glue packages are not supported
  ```julia
  # This symbol is only defined on Julia versions that support glue packages

>   if isdefined(Base, :get_gluepkg)

Reviewer 3: ```suggestion
  if !isdefined(Base, :get_gluepkg)
```

If I understand correctly?

Author: Yes, thanks :)

Every dependency should in general have a compatibility constraint on it.
This is an important topic so there is a separate chapter about it: [Compatibility](@ref Compatibility).

## Weak dependencies

!!! note
    This is a somewhat advanced usage of Pkg which can be skipped for people new to Julia and Julia packages.

A weak dependency is a dependency that will not automatically install when the package is installed but

> you can still control what versions of that package is allowed to install by setting compatibility on it.

Reviewer: ```suggestion
you can still control what versions of that package are allowed to be installed by setting compatibility on it.
```

An extension will only be loaded if the extension dependencies are loaded from the same environment or environments higher in the environment stack than the package itself.

!!! compat
    Often you will put the extension dependencies into the `test` target so they are loaded when running e.g. `Pkg.test()`. On earlier Julia versions
    this requires you to also put the package in the `[extras]` section. This is unfortunate but the project verifier on older Julia versions will
    complain if this is not done.

!!! note
    If you use a manifest generated by a Julia version that does not know about extensions with a Julia version that does
    know about them, the extensions will not load. This is because the manifest lacks some information that tells Julia
    when it should load these packages. So make sure you use a manifest generated at least the Julia version you are using.

### Backwards compatibility

This section discusses various methods for using extensions on Julia versions that support them,
while simultaneously providing similar functionality on older Julia versions.
#### Requires.jl

> This section is relevant if you are currently using Requries.jl but want to transition to using extensions (while still having Requires be used on Julia versions that do not support extensions).

Reviewer: ```suggestion
This section is relevant if you are currently using Requires.jl but want to transition to using extensions (while still having Requires be used on Julia versions that do not support extensions).
```

This is done by making the following changes (using the example above):

- Add the following to the package file. This makes it so that Requires.jl loads and inserts the
  callback only when extensions are not supported
  ```julia
  # This symbol is only defined on Julia versions that support extensions
  if !isdefined(Base, :get_extension)
  using Requires
  function __init__()

>       @require Contour = "d38c429a-6771-53c6-b99e-75d170b6e991" include("../ext/ContourExt.jl)

Reviewer: ```suggestion
      @require Contour = "d38c429a-6771-53c6-b99e-75d170b6e991" include("../ext/ContourExt.jl")
```
